Implement beginless range

LillianZ opened the pull request:
Starting to implement beginless ranges for Ruby 2.7 compatibility https://github.com/oracle/truffleruby/issues/2004

So far:
Implemented or added specs for Range#{new, bsearch, count, each, equal_value, first, inspect, max, min, size} regarding beginless range compatibility
Implemented or added specs for Array#{[], []=, slice, slice!, to_a, fill, values_at} regarding beginless range compatibility

To do:
Range#cover?/include/===
some or all of Range#{entires, eql?, hash, last, member?, minmax, step/%, to_s}
Methods outside of Array and Range, such as String methods.

https://github.com/Shopify/truffleruby/issues/1

eregon commented:
@norswap Could you review this one?

Looks good to me :)

LillianZ commented:
Is `jt format` a known issue right now?

norswap commented:
@LillianZ Do you mean the Eclipse error where it can't find `libjvm.dylib` on macOS? Then yes. The quickfix is to:

```
OPENJDK_VERSION=openjdk1.8.0_272-jvmci-20.3-b04 # adapt
cd ~/.mx/cache/truffleruby/$OPENJDK_VERSION/Contents/MacOS/
rm libjli.dylib
ln -s ../Home/jre/lib/jli/libjli.dylib libjli.dylib
cd -
```

The problem has been known for some time, but it's in limbo as it requires digging into the OpenJDK build process. It's fixed on the GraalVM image builds however.

LillianZ commented:
Got it, thank you! I misunderstood the error logs.

norswap commented on src/main/ruby/truffleruby/core/range.rb:
```diff
@@ -344,7 +344,7 @@ def ===(value)
 
   def inspect
-    result = "#{self.begin.inspect}#{exclude_end? ? "..." : ".."}#{Truffle::RangeOperations.endless?(self) ? "" : self.end.inspect}"
+    result = "#{(self.begin == nil && self.end != nil) ? "" : self.begin.inspect}#{exclude_end? ? "..." : ".."}#{(Truffle::RangeOperations.endless?(self) && self.begin != nil) ? "" : self.end.inspect}"
```
We should probably have `RangeOperations.beginless?` for symmetry. Since it's a `nil` check, it should probably be `Primitive.range_endless/beginless` ultimately. Very minor & can be taken care of later.

LillianZ replied:
I can definitely add a `RangeOperations.beginless?` method, although if a range is beginless and endless `(nil..nil)` then in MRI both sides get printed, so the other end need to be checked anyways.

I personally would find it more visually appealing in this case to use neither `beginless?` nor `endless?` (unless the check on the other end happens in `beginless?` or `endless?`, but redefining `endless?` could break things).

norswap replied:
Good point, what do think of:

```ruby
def inspect
  sep = exclude_end? ? "..." : ".."
  if (Primitive.nil? self.begin && Primitive.nil? self.end)
    result = "nil#{sep}nil"
  else
    result = (Primitive.nil? self.begin ? "" : self.begin.inspect) + sep +
             (Primitive.nil? self.end ? "" : self.end.inspect)
  end
  Primitive.infect(result, self)
end
```

LillianZ replied:
I love it. 

I'll commit it to the PR and squash it with the original Range#inspect when I rearrange the the commits after I'm done?

On that note, is there a preferred organization? Such as: All specs first; or Range behaviour first, Array behaviour follows, etc; or specs precede relevant code changes; or it really doesn't matter.

eregon replied:
The order of commits doesn't matter for specs.

norswap reviewed (commented):
Looks good to me so far!

Another place I remember I had to handle endless ranges explicitly was `StringNodes#GetIndexNode`.

eregon commented:
@LillianZ I can add that email to the list, but it would be better to rebase instead of merge to have a clean history.

LillianZ commented:
Sorry, a conflict appeared online not appearing locally, and trying to resolve it caused a merge. I have been trying to rebase.

eregon commented:
FWIW, the GitHub UI often thinks there are conflicts in CHANGELOG.md, but actually there isn't due to the union merge strategy which never conflicts. Best is to just ignore those, because there are no actual conflicts.

LillianZ commented:
To confirm: it's okay when GitHub UI complains about a conflict (at least in CHANGELOG) because it's something you ca fix?

LillianZ commented:
Also, while working on Range#bsearch I fixed some specs involving infinity bounds so that the blocks are appropriately monotonic false ... true or 1 ... 0 ... -1 . Is it okay that I included them here, or should be elsewhere, since they are not related to beginless ranges?

eregon commented:
> To confirm: it's okay ignore when GitHub UI complains about a conflict (at least in CHANGELOG) because it's something you can fix?

Yes. And actually there is nothing needed to fix it, it's a bug of whatever GitHub uses to detect conflicts.

eregon commented:
It's still fairly related so that's fine.

LillianZ commented:
Before the diff gets too large, it could make sense to just merge this PR and finish on another?

So far:
Implemented or added specs for Range#{new, bsearch, count, each, equal_value, first, inspect, max, min, size}
Implemented or added specs for Array#{[], []=, slice, slice!, to_a, fill, values_at}

To do:
Range#cover?/include/===
some or all of Range#{entires, eql?, hash, member?, minmax, step, to_s}
Methods outside of Array and Range, such as String methods.

eregon commented:
> Before the diff gets too large, it could make sense to just merge this PR and finish on another?

Yes, definitely, I was actually about to suggest that.

norswap commented on spec/ruby/core/array/slice_spec.rb:
```diff
@@ -165,6 +165,18 @@ def to.to_int() -2 end
+  ruby_version_is "2.7" do
+    it "works with beginless ranges" do
+      a = [0,1,2,3,4]
+      a.slice!(Range.new(nil, 3)).should == [0, 1, 2, 3]
```
minor: could use dot notation for consistency (other places too)

LillianZ replied:
done! I kept on running into CI issues using dot notation, but I now realize I probably missed an eval somewhere.

norswap commented on spec/ruby/core/range/bsearch_spec.rb:
```diff
@@ -126,8 +126,8 @@
-        (-inf...0).bsearch { |x| x == -inf }.should == nil
+        (-inf..0).bsearch { |x| x != -inf }.should == -Float::MAX
+        (-inf...0).bsearch { |x| x != -inf }.should == -Float::MAX
         (inf..inf).bsearch { |x| true }.should == inf
```
I think those are actually correct, and should be kept (in addition to the new ones). Ditto for those below in this file.

LillianZ replied:
For bsearch the block must be either monotonically false ... true or monotonically 1 ... 0 ... -1 . (https://ruby-doc.org/core-2.7.2/Range.html#method-i-bsearch explains it better than me). The blocks I replaced were monotonically the other way around, like in this case -inf is true and everything larger is false; so I just flipped the values.

norswap replied:
Forgot to answer this. Okay, makes sense. I was coming from the perspective that these tests work on MRI, but I think it's fair that nobody is going to rely on this.

norswap commented on src/main/java/org/truffleruby/core/range/RubyObjectRange.java:
```diff
@@ -29,12 +29,20 @@ public RubyObjectRange(RubyClass rubyClass, Shape shape, boolean excludedEnd, Ob
     }
 
+    public boolean isNilNil() {
+        return begin == Nil.INSTANCE && end == Nil.INSTANCE;
```
Those should be `Primitive.nil?`

LillianZ replied:
I haven't figured out how to use Primitives in Java code; could you point me to some part of the codebase where this happens or rewrite this function for me? Thanks!

norswap replied:
Sorry, this is just me being absent-minded and thinking this was Ruby. It's actually perfectly fine as-is!

Primitives are just a means to avoid method call overhead in Ruby code. They're just Truffle nodes with the `@Primitive` annotation. They can be reused like other nodes in Java (e.g. via `@Cached MyPrimitiveNode` in the parameter list of a method marked `@Specialization`) - but this is just useful to reuse logic, there is no performance gain from using them in Java code.

You can find the code for a given primitive in the generated stubs (e.g. for `Primitive.nil?`, look in `src/main/java/org/truffleruby/core/support/TypeNodes.java`, there's an auto-generated comment pointing at `org.truffleruby.core.support.TypeNodes.IsNilNode`).

For an example of primitive being reused, see for instance `ArrayIndexNodes.ReadNormalizedNode`.

LillianZ replied:
Oh awesome example, thanks!!!

norswap commented on src/main/ruby/truffleruby/core/range.rb:
```diff
@@ -97,7 +99,7 @@ def bsearch(&block)
   private def bsearch_float(&block)
-    normalized_begin = self.begin.to_f
+    normalized_begin = Primitive.nil?(self.begin) ? -Float::INFINITY : self.begin.to_f
     normalized_end = Truffle::RangeOperations.endless?(self) ? Float::INFINITY : self.end.to_f
```
Mixed `Primitive.nil?` and `TruffleRangeOperations.endless?`.
I think we should just get rid of `TruffleRangeOperations.endless?/beginless?` and just use `Primitive.nil?` everywhere.

I'm not entirely certain why `endless?` exists in the first place, but I _think_ it was because it used to be a primitive, which was then redefined in terms of `Primitive.nil?`. No point in keeping it now.

norswap commented on spec/ruby/core/range/count_spec.rb:
```diff
@@ -0,0 +1,20 @@
+      inf = Float::INFINITY
+      (..."p").count.should == inf
+      (...nil).count.should == inf
+      (..10.0).count.should == inf
```
Should use `eval` so that it will be able to parse on older Ruby versions. In other places too, both for beginless and endless ranges.

eregon replied:
The GH Actions check should catch that, but it unfortunately fails since yesterday. Should be fixed soon.

eregon replied:
Should be fixed now, any new commit/push should trigger the usual actions.

LillianZ replied:
That is so useful know. GH actions is also so useful; there were so many bugs that I snuck through while it was off. Sorry!

norswap commented on spec/ruby/core/range/size_spec.rb:
```diff
@@ -25,10 +25,24 @@
+    end
+  end
+
+  ruby_version_is "2.6" do
```
should be 2.7 :)

norswap commented on src/main/ruby/truffleruby/core/range.rb:
```diff
@@ -262,6 +262,16 @@ def bsearch(&block)
+      return Float::INFINITY unless block_given? || !Primitive.undefined?(item)
+    end
+
+    p "count for here"
```
woops

norswap reviewed (commented):
Great work 👍

LillianZ commented:
I have am slightly concerned that 2.6 specs have been running for 3 hours, but they passed locally so I'm not sure what to do.

chrisseaton commented:
Looks broken - it's not running at all. CI is hard.

LillianZ commented:
Thanks for checking! I fixed my commit message to get CI to rerun.

eregon commented:
Most likely this indicates a spec getting stuck when run on MRI 2.6.6.
Unfortunately it seems GitHub Actions stops the job without any signal or anything to get a chance to show a backtrace.
Locally I could reproduce with:
```
$ mspec --timeout 60 .
$ ruby /home/eregon/code/mspec/bin/mspec-run --timeout 60 .
ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
[\ | ===============   39%                    | 00:00:33]      0F      0E 
Range#count returns Infinity for endless ranges without arguments or blocks
Example took longer than the configured timeout of 60.0s
```

So I'll add the `--timeout` option, that way we'll see which spec got stuck in CI.

chrisseaton commented:
Looks like it's generating some kind of infinite string or something. We're investigating.

LillianZ commented:
It's fixed! For some reason I didn't realize the specs breaking on CRuby 2.6 => fix 2.6 specs.

eregon commented on spec/ruby/core/array/fill_spec.rb:
```diff
@@ -324,4 +324,11 @@ def obj.<=>(rhs); rhs == self ? 0 : nil end
+  ruby_version_is "2.7" do
+    it "works with beginless ranges" do
+      [1, 2, 3, 4].fill('x', eval("(nil..2)")).should == ["x", "x", "x", 4]
+      [1, 2, 3, 4].fill(eval("(nil...2)")) { |x| x + 2 }.should == [2, 3, 3, 4]
```
Could you remove the extra `nil` here? Also happens in other specs.

norswap commented:
Failures:
- `Primitive.infect` was removed as taint checking is deprecated in 2.7 and the various tainting methods become no-ops
- The `Range#max` spec is tagged for Ruby 2.7.2, but the fix to that upstream bug is [here](https://github.com/ruby/ruby/commit/8a5ad2b77d7a24e4f8f4fef179ae5efced935f91) and it seems to be on a branch tagged for 3.0.

LillianZ commented:
Woot! everything's finally passing! sorry for the delay, and thanks for all the tips; I found the range#max patch in ruby-lang and definitely would not have known to go looking in github!

norswap commented:
Merged! Congrats 👍
